FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RecipientsViewModel.kt

```diff
@@ -0,0 +1,391 @@
+          Toast.makeText(
+            context,
+            context.getString(R.string.could_not_save_new_recipient), Toast.LENGTH_LONG
+          ).show()
+        }
+      }
+    }
+  }
+
+  fun copyPubKeysToRecipient(recipientEntity: RecipientEntity?, pgpKeyDetails: PgpKeyDetails) {
+    viewModelScope.launch {
+      recipientEntity?.let {
+        val existingPubKey = roomDatabase.pubKeyDao()
+          .getPublicKeyByRecipientAndFingerprint(recipientEntity.email, pgpKeyDetails.fingerprint)
+        if (existingPubKey == null) {
+          roomDatabase.pubKeyDao()
+            .insertSuspend(pgpKeyDetails.toPublicKeyEntity(recipientEntity.email))
+        }
+      }
+    }
+  }
+
+  fun updateExistingPubKey(publicKeyEntity: PublicKeyEntity, pgpKeyDetails: PgpKeyDetails) {
+    viewModelScope.launch {
+      roomDatabase.pubKeyDao()
+        .updateSuspend(
+          pgpKeyDetails.toPublicKeyEntity(publicKeyEntity.recipient).copy(
+            id = publicKeyEntity.id,
+            recipient = publicKeyEntity.recipient
+          )
+        )
+    }
+  }
+
+  fun copyPubKeysBetweenRecipients(
+    sourceRecipientEntity: RecipientEntity?,
+    destinationRecipientEntity: RecipientEntity?
+  ) {
+    viewModelScope.launch {
+      sourceRecipientEntity ?: return@launch
+      destinationRecipientEntity ?: return@launch
+
+      val existingPubKeysOfSource =
+        roomDatabase.pubKeyDao().getPublicKeysByRecipient(sourceRecipientEntity.email)
+
+      if (existingPubKeysOfSource.isNotEmpty()) {
+        for (existingPubKey in existingPubKeysOfSource) {
+          val fetchedPubKey = roomDatabase.pubKeyDao().getPublicKeyByRecipientAndFingerprint(
+            destinationRecipientEntity.email,
+            existingPubKey.fingerprint
+          )
+
+          if (fetchedPubKey == null) {
+            roomDatabase.pubKeyDao()
+              .insertSuspend(
+                existingPubKey.copy(
+                  id = null,
+                  recipient = destinationRecipientEntity.email.lowercase()
+                )
+              )
+          }
+        }
+      }
+    }
+  }
+
+  fun filterContacts(searchPattern: String?) {
+    searchPatternLiveData.value = searchPattern
+  }
+
+  fun deleteContactByEmail(email: String) {
+    viewModelScope.launch {
+      roomDatabase.recipientDao().getRecipientByEmailSuspend(email)?.let {
+        roomDatabase.recipientDao().deleteSuspend(it)
+      }
+    }
+  }
+
+  fun fetchPubKeys(keyIdOrEmail: String) {
+    viewModelScope.launch {
+      lookUpPubKeysMutableStateFlow.value = Result.loading()
+      val activeAccount = getActiveAccountSuspend()
+      lookUpPubKeysMutableStateFlow.value =
+        controlledRunnerForPubKeysFromServer.cancelPreviousThenRun {
+          return@cancelPreviousThenRun apiRepository.pubLookup(
+            context = getApplication(),
+            identData = keyIdOrEmail,
+            orgRules = activeAccount?.clientConfiguration
+          )
+        }
+    }
+  }
```

Comment:
Isn't this the same as `getPgpInfoFromServer`? If it has more or less the same purpose, should call it similar to that method but differentiate the purpose in the name. So `getPublicKeysFromRemoteServersXxxxx` 

Also, `keyIdOrEmail` should just be `email`.

Reply:
> So getPublicKeysFromRemoteServersXxxxx

I think we can do that